### PR TITLE
Add section on extending napari

### LINKED ISSUE #239: Add `magicgui.widgets.Widget` subclass widget example
## 📚 New content request
Add `magicgui.widgets.Widget` subclass widget example to the cookiecutter and the building a plugin [widgets section](https://napari.org/dev/plugins/building_a_plugin/guides.html#widgets).

Currently there are not examples of using `magicgui.widgets.Widget` which @DragaDoncila says is less complex than using `QWidget`.

### LINKED ISSUE #244: Update `magicgui` guide for `npe2`
## 🧰 Task
The "Using magicgui in napari plugin widgets" [guide](https://napari.org/stable/guides/magicgui.html#using-magicgui-in-napari-plugin-widgets) has lots of fantastic information, but unfortunately ends with totally outdated info about its use in plugin widgets. I think we should either remove that section entirely and just link to the plugin guides, or better yet, update it for `npe2`. We should also likely be linking to it from the widget contribution guide.

### PR DESCRIPTION
**Pull request by lucyleeow**
# References and relevant issues
closes #244
closes #239

# Description
This was originally meant to be a small PR closing #244, but has been extended to adding an 'extending napari' section (as it was difficult to break it up to smaller PRs), loosely based off of @DragaDoncila slides from [here](https://github.com/napari/docs/issues/244#issuecomment-1831360814).

* Add 'Extending napari' section under 'How-to guides' - covers widgets, ~intro to event loop~ and how to connect your how events
* "Hooking up your own events" and ~"An introduction to the event loop in napari"~ pages mostly unchanged, just moved (edit: 'An introduction to the event loop in napari' will remain where it is, see: https://github.com/napari/docs/pull/286#discussion_r1418490112)
* Big changes to the "Using magicgui in napari" page:
  * This is now documents how to build widgets with magicgui and by creating a widget class. Also describes how to add widget to the viewer and how to adapt the widget to be a plugin widget contribution. Adds more extensive intro.
  * Adds lots of cross references to magicgui, the plugins section and the cookie-cutter

### CODE REVIEW

**Comment by lucyleeow** [comment]
cc @DragaDoncila

**Comment by lucyleeow** [comment]
Okay I think this is ready for review now. This was difficult to write because I am not well versed in Qt stuff. Happy to make changes, as I may have mis-understood things.

I was not sure what the best link to use for `QtWidgets.QWidget` was. @DragaDoncila said https://doc.qt.io/qt-5/qwidget.html is fine (but does it matter qt 5 or qt 6?!?)

I will need to update the plugins widget contributions page and add cross reference to this page (done in https://github.com/napari/npe2/pull/331). This is in `npe2` repo though (3 repo docs :unamused:  ) so will be done in another PR.

File: docs/howtos/extending/magicgui.md [hunk 1]
```diff
@@ -12,17 +12,58 @@ kernelspec:
...
+* via {meth}`napari.qt.Window.add_dock_widget` in a python script or interactive
+  console (see [How to launch napari](getting_started) for details on launching
+  and interacting programmatically with `napari`)
+* adding a [widget contribution](widgets-contribution-guide) in a
+  [plugin](plugins-index).
+
+There is an important implementation distinction between the two methods;
+{meth}`napari.qt.Window.add_dock_widget` expects an *instance* of a widget, like a
+{class}`magicgui.widgets.Widget` or a {class}`qtpy.QtWidgets.QWidget`, whereas
+[Widget contributions](widgets-contribution-guide), expects a widget *class*
```
Comment (DragaDoncila) [inline docs/howtos/extending/magicgui.md]:
```suggestion
[Widget contributions](widgets-contribution-guide), expect a `callable` (like a function or a class) that will return a widget instance.
```

File: docs/howtos/extending/magicgui.md [hunk 2]
```diff
@@ -526,15 +615,46 @@ widget2 = my_factory(call_button=False, x={'widget_type': 'Slider'})
...
+We will begin with the simplest but least extensible parent class and end with the
+parent class the most extensible.
+
 ### `magicgui.widgets.FunctionGui`
 
-The other option for using `magicgui` in plugins is to directly subclass
-{class}`magicgui.widgets.FunctionGui` (which is the type that is returned
-by the {func}`@magicgui <magicgui.magicgui>` decorator).
+Creating a widget by subclassing {class}`magicgui.widgets.FunctionGui` is similar in
+principal to using the {func}`@magicgui <magicgui.magicgui>` decorator. Decorating
```
Comment (DragaDoncila) [inline docs/howtos/extending/magicgui.md]:
```suggestion
principle to using the {func}`@magicgui <magicgui.magicgui>` decorator. Decorating
```

File: docs/howtos/extending/magicgui.md [hunk 3]
```diff
@@ -526,15 +615,46 @@ widget2 = my_factory(call_button=False, x={'widget_type': 'Slider'})
...
+a function with {func}`@magicgui <magicgui.magicgui>` is equivalent to passing
+the same function to {class}`magicgui.widgets.FunctionGui`'s `function` parameter.
+The remaining {class}`magicgui.widgets.FunctionGui` parameters essentially
+mirror {func}`@magicgui <magicgui.magicgui>`'s parameters.
+Indeed, {class}`magicgui.widgets.FunctionGui` is the type that is returned by
+{func}`@magicgui <magicgui.magicgui>`. Subclassing
+{class}`magicgui.widgets.FunctionGui` however, gives you access to the
+`native` `QWidget` of your widget, allowing you change its appearance and add
```
Comment (DragaDoncila) [inline docs/howtos/extending/magicgui.md:663]: Can `FunctionGui` guis receive the viewer as a parameter?

File: docs/howtos/extending/index.md [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+
+There are a number of ways to extend your usage of `napari`:
+
+* [**Adding widgets**](creating-widgets): adding a widget (new GUI interaction element)
+  to your `napari` app.
+* **Connecting callbacks**: customize the behavior of `napari` by connecting an action
+  (function) to an [event](intro-to-event-loop) such as:
+  - [**Key press**](connect-key-event): perform an action based on a keyboard shortcut
+  - [**Mouse event**](connect-mouse-event): perform an action based on a mouse event
+  - [**Layer/viewer event**](connect-napari-event): perform an action when a layer or
```
Comment (DragaDoncila) [inline docs/howtos/extending/index.md:12]: I kinda want to split responding to `Layer/viewer events` with custom callbacks and adding your own key press/mouse event handlers, but maybe there's not enough distinction? With key preses/mouse events you're defining both the event and the callback, so you could be introducing new events even. While responding to layer/viewer events is exactly that: you're hooking into existing events emitted by the viewer to perform custom actions. But maybe it's not really worth distinguishing idk

File: docs/howtos/extending/magicgui.md [hunk 5]
```diff
@@ -48,44 +89,89 @@ def widget_demo(
...
+
+```{note}
+For a more complex example of a `magicgui.magicgui` widget, see the
+[gaussian blur example](https://pyapp-kit.github.io/magicgui/generated_examples/napari/napari_parameter_sweep/#napari-parameter-sweeps)
+in the `magicgui` documentation.
+```
+
+(magicgui-parameter-annotations)=
+
 ## Parameter annotations
```
Comment (DragaDoncila) [inline docs/howtos/extending/magicgui.md]:
```suggestion
### Parameter annotations
```

Should this be a triple heading to fit under the magicgui bit? And same with `Return annotations` below?

File: docs/howtos/extending/magicgui.md [hunk 6]
```diff
@@ -48,44 +89,89 @@ def widget_demo(
...
+viewer = napari.view_image(data.camera())
+# Add widget to viewer
+viewer.window.add_dock_widget(threshold_magic_widget)
 ```
 
+```{code-cell} python
+:tags: [remove-input]
+from napari.utils import nbscreenshot
+
+viewer.window._qt_window.resize(750, 550)
```
Comment (DragaDoncila) [inline docs/howtos/extending/magicgui.md]: could we maybe make this a bit bigger? The screenshots really squish up the image. If we can't I think it's nbd

File: docs/howtos/extending/event_loop.md [hunk 7]
```diff
@@ -1,20 +1,47 @@
 (intro-to-event-loop)=
 
-# An introduction to the event loop in napari
```
Comment (DragaDoncila) [inline docs/howtos/extending/event_loop.md:3]: Not sure if this doc fits in this section? It's not really a guide to extending napari at all. I would maybe keep this under `in depth explanations` and link to it from the top of the `Hooking up your own events` doc?

File: docs/howtos/extending/connecting_events.md [hunk 8]
```diff
@@ -88,6 +84,8 @@ and [mouse drag
 callback](https://github.com/napari/napari/blob/main/examples/mouse_drag_callback.py)
 examples.
 
+(connect-napari-event)=
+
 ## Connection functions to native napari events
```
Comment (DragaDoncila) [inline docs/howtos/extending/connecting_events.md]:
```suggestion
## Connecting functions to native napari events
```

File: docs/howtos/extending/index.md (hunk 4 above)
Comment (DragaDoncila) [reply docs/howtos/extending/index.md:12]: I see that those are all in the same doc anyway so maybe just ignore this

File: docs/howtos/extending/magicgui.md [hunk 9]
```diff
@@ -548,7 +668,103 @@ class MyGui(FunctionGui):
...
+
+Class widgets are easy to use as
+[plugin widget contributions](widgets-contribution-guide). Simply provide the
+class definition and add to the plugin manifest.
+
+### `magicgui.widgets.Container`
+
+The {class}`magicgui.widgets.Container` allows you to build more complex widgets
+from sub-widgets. This gives you more control over each sub-widget and how callbacks
+are connected to events.
```
Comment (DragaDoncila) [inline docs/howtos/extending/magicgui.md]:
```suggestion
are connected to events. You can still use the convenient `magicgui` widget generation features as shown below.
```

**Review by DragaDoncila — COMMENTED** [review]
@lucyleeow left a couple of comments! Looks great ❤️

File: docs/howtos/extending/magicgui.md (hunk 1 above)
Comment (lucyleeow) [reply docs/howtos/extending/magicgui.md]: Agree, i kept the original wording but it felt wrong to me too.

File: docs/howtos/extending/magicgui.md (hunk 3 above)
Comment (lucyleeow) [reply docs/howtos/extending/magicgui.md:663]: I would think so but let me check

File: docs/howtos/extending/magicgui.md (hunk 5 above)
Comment (lucyleeow) [reply docs/howtos/extending/magicgui.md]: Yes I changed this then changed it back. But you're right,  you don't annotate a widget class in the same way

File: docs/howtos/extending/event_loop.md (hunk 7 above)
Comment (lucyleeow) [reply docs/howtos/extending/event_loop.md:3]: Yeah good point, will amend

File: docs/howtos/extending/magicgui.md (hunk 6 above)
Comment (lucyleeow) [reply docs/howtos/extending/magicgui.md]: Done, hopefully better.

File: docs/howtos/extending/magicgui.md (hunk 3 above)
Comment (lucyleeow) [reply docs/howtos/extending/magicgui.md:663]: Checked with the plugins PR (#4991) and yes it can. Does that affect it accessing the native `QWidget` ?!

Comment (DragaDoncila) [reply docs/howtos/extending/magicgui.md:663]: nope, doesn't affect that, I was just curious. I think a big benefit of the "class based" vs. "function based" widgets is that you can access the viewer and connect more complex events/callbacks in a simpler and more direct way. Function based widgets can also receive the viewer and do all that stuff, but the code ends up being more spaghettified

Comment (lucyleeow) [reply docs/howtos/extending/magicgui.md:663]: I think magicgui functions are able to access the viewer! Its done via `magicgui.register_type`

Comment (DragaDoncila) [reply docs/howtos/extending/magicgui.md:663]: They are indeed. I just find when looking at plugin code that there's a correlation between function based widgets that take the viewer and needlessly complex logical flow - I suspect in many of these cases using a class based design would simplify things. But that's kinda hard to capture in this one doc lol so you can mostly ignore this comment chain for the purposes of review

Comment (lucyleeow) [reply docs/howtos/extending/magicgui.md:663]: >  plugin code that there's a correlation between function based widgets that take the viewer and needlessly complex logical flow

I had no idea, why is it more complicated to use the viewer via a function?

File: docs/guides/event_loop.md [hunk 10]
```diff
@@ -81,10 +102,10 @@ while True:  # infinite loop!
             process_event(event)
 ```
 
-Actions taken by the user add events to the queue (e.g. "button pressed",
-"slider moved", etc...), and the event loop handles them one at a time.
+Actions taken by the user add events to the queue and the event loop handles them one
+at a time.
 
-## Qt applications and event loops
+## Qt applications, event loops and widgets
```
Comment (psobolewskiPhD) [inline docs/guides/event_loop.md]:
```suggestion
## Qt applications, event loops, and widgets
```

File: docs/guides/event_loop.md [hunk 11]
```diff
@@ -1,20 +1,47 @@
 (intro-to-event-loop)=
```
Comment (psobolewskiPhD) [inline docs/guides/event_loop.md:1]: Why isn't this file under `howtos/extending/` like the magicgui one?
(going by the _toc.yml)
Doesn't show up in the html toc:
image

Seems like this page is only a nested link within the `extending` index.md

**Review by psobolewskiPhD — CHANGES_REQUESTED** [review]
Love this!!
❤️ 
Small nitpick plus a comment. Maybe I'm confused...

File: docs/guides/event_loop.md (hunk 11 above)
Comment (lucyleeow) [reply docs/guides/event_loop.md:1]: Sorry, see: https://github.com/napari/docs/pull/286#discussion_r1418490112 @DragaDoncila suggested that this remain where it is. I will update my OP to reflect

Comment (psobolewskiPhD) [reply docs/guides/event_loop.md:1]: That's fine, then just put it back in the _toc.yml

Comment (lucyleeow) [reply docs/guides/event_loop.md:1]: Ah whoops, forgot to update the toc! Good pick up!

File: docs/howtos/extending/magicgui.md [hunk 12]
```diff
@@ -0,0 +1,784 @@
...
+  language: python
+  name: python3
+---
+(creating-widgets)=
+
+# Creating widgets
+
+Widgets are small composable graphical elements that can be added to the `napari` user
+interface. The easiest way to add a widget is by using
+[`magicgui`](https://pyapp-kit.github.io/magicgui/), a python package that assists
```
Comment (psobolewskiPhD) [inline docs/howtos/extending/magicgui.md]:
```suggestion
[`magicgui`](https://pyapp-kit.github.io/magicgui/), a Python package that assists
```

File: docs/howtos/extending/magicgui.md [hunk 13]
```diff
@@ -0,0 +1,784 @@
...
+in building widgets. It is a general abstraction layer on GUI toolkit backends (like
+Qt), with an emphasis on mapping python types to widgets. This enables you to easily
```
Comment (psobolewskiPhD) [inline docs/howtos/extending/magicgui.md]:
```suggestion
Qt), with an emphasis on mapping Python types to widgets. This enables you to easily
```

File: docs/howtos/extending/magicgui.md [hunk 14]
```diff
@@ -0,0 +1,784 @@
...
+   [`magicgui`](https://pyapp-kit.github.io/magicgui/) (simplest but least extensible
+   and flexible)
+2. create a widget class that subclasses a
+   [`magicgui` widget class](https://pyapp-kit.github.io/magicgui/widgets/#the-widget-hierarchy)
+3. create a widget class that subclasses
+   [`QtWidgets.QWidget`](https://doc.qt.io/qt-5/qwidget.html) (most extensible but also
+   the most difficult to implement)
+
+For more examples, the [GUI gallery examples](https://napari.org/stable/_tags/gui.html)
+has some widget examples. Additionally,
```
Comment (psobolewskiPhD) [inline docs/howtos/extending/magicgui.md]:
```suggestion
For more examples, see the [GUI gallery examples](https://napari.org/stable/_tags/gui.html). Additionally,
```

File: docs/howtos/extending/magicgui.md [hunk 15]
```diff
@@ -0,0 +1,784 @@
...
+[cookiecutter-napari-plugin](https://github.com/napari/cookiecutter-napari-plugin)
+has more robust widget examples that you can adapt to your needs.
+
+There are two ways to then add a widget to a `napari` viewer:
+
+* via {meth}`napari.qt.Window.add_dock_widget` in a python script or interactive
```
Comment (psobolewskiPhD) [inline docs/howtos/extending/magicgui.md]:
```suggestion
* via {meth}`napari.qt.Window.add_dock_widget` in a Python script or interactive
```

File: docs/howtos/extending/magicgui.md [hunk 16]
```diff
@@ -0,0 +1,784 @@
...
+  console (see [How to launch napari](getting_started) for details on launching
+  and interacting programmatically with `napari`)
+* adding a [widget contribution](widgets-contribution-guide) in a
```
Comment (psobolewskiPhD) [inline docs/howtos/extending/magicgui.md]:
```suggestion
* by adding a [widget contribution](widgets-contribution-guide) in a
```

File: docs/howtos/extending/magicgui.md [hunk 17]
```diff
@@ -0,0 +1,784 @@
...
+For a more complex example of a `magicgui.magicgui` widget, see the
+[gaussian blur example](https://pyapp-kit.github.io/magicgui/generated_examples/napari/napari_parameter_sweep/#napari-parameter-sweeps)
+in the `magicgui` documentation.
+```
+
+(magicgui-parameter-annotations)=
+
+### Parameter annotations
+
+The following napari types may be used as *parameter* type annotations in
```
Comment (psobolewskiPhD) [inline docs/howtos/extending/magicgui.md:170]:
```suggestion
The following `napari` types may be used as *parameter* type annotations in
```

File: docs/howtos/extending/magicgui.md [hunk 18]
```diff
@@ -0,0 +1,784 @@
...
+If you find that you require the viewer instance because of functionality that
+is otherwise missing here, please consider opening an issue in the
+[napari issue tracker](https://github.com/napari/napari/issues/new/choose),
+describing your use case.
+```
+(magicgui-return-annotations)=
+
+### Return annotations
+
+The following napari types may be used as *return* type annotations in `magicgui`
```
Comment (psobolewskiPhD) [inline docs/howtos/extending/magicgui.md:299]:
```suggestion
The following `napari` types may be used as *return* type annotations in `magicgui`
```

File: docs/howtos/extending/magicgui.md [hunk 19]
```diff
@@ -0,0 +1,784 @@
...
+```python
+@magicgui
+def my_func(data: 'napari.types.ImageData') -> 'napari.types.ImageData':
+    ...
+```
+
+:::{tip}
+
+If you'd like to maintain IDE type support and autocompletion, you can
+do so by hiding the napari imports inside of a {attr}`typing.TYPE_CHECKING`
```
Comment (psobolewskiPhD) [inline docs/howtos/extending/magicgui.md:558]:
```suggestion
do so by hiding the `napari` imports inside of a {attr}`typing.TYPE_CHECKING`
```

**Comment by psobolewskiPhD** [comment]
Sorry a few more nitpicks!

File: docs/howtos/extending/magicgui.md (hunk 14 above)
Comment (lucyleeow) [reply docs/howtos/extending/magicgui.md]: This is awkward because not all of the 'GUI' examples are widget examples. I've added this in parenthesis.

**Comment by lucyleeow** [comment]
Okay I think everything has been addressed! I've gone and fixed 'Python' and 'napari' in the files touched here.

**Review by psobolewskiPhD — APPROVED** [review]

**Review by DragaDoncila — APPROVED** [review]
Sorry for holding this one up. Thanks @lucyleeow :pray:

**Comment by lucyleeow** [comment]
@psobolewskiPhD @DragaDoncila could this be merged to prevent difficult merge conflicts? Thanks

**Comment by psobolewskiPhD** [comment]
Sorry, overlooked it--yes I'll merge once CI finishes.

**Pull request merged**